File: scripts/refresh.py
#!/usr/bin/env python3
"""
Bambrew Grants Dashboard — News Refresher
==========================================
Fetches the latest Google News results for Bambrew-relevant keywords and
rewrites the NEWS array in ~/bambrew-grants-dashboard.html in place.

- Reads keywords from keywords.txt (one per line)
- Uses Google News RSS — no API key required
- Stdlib only (no pip install)
- Writes a backup of the dashboard before every change
- Logs every run to refresh.log

Run manually:    python3 refresh.py
Scheduled:       installed via install.sh (launchd, every Monday 9am)
"""

from urllib.parse import urlencode
from urllib.request import urlopen, Request
from urllib.error import URLError
from email.utils import parsedate_to_datetime
import xml.etree.ElementTree as ET
import datetime
import html
import logging
import pathlib
import ssl
import sys

# macOS Python (Homebrew + python.org installers) often lacks a configured CA
# bundle, which makes urlopen() fail with CERTIFICATE_VERIFY_FAILED on HTTPS.
# We're reading public RSS only (no creds sent), so we try verified first and
# transparently fall back to an unverified context on cert errors.
_VERIFIED_CTX = ssl.create_default_context()
_UNVERIFIED_CTX = ssl._create_unverified_context()

# ---------------------------------------------------------------
# Config
# ---------------------------------------------------------------
HERE          = pathlib.Path(__file__).parent.resolve()
DASHBOARD     = (HERE.parent / "index.html").resolve()
KEYWORDS_FILE = HERE / "keywords.txt"
LOG_FILE      = HERE / "refresh.log"

MAX_ITEMS_TOTAL   = 12   # cards shown in the What's New drawer
PER_KEYWORD_MAX   = 2    # cap per keyword (prevents one keyword dominating)
DAYS_BACK         = 30   # only keep stories from the last N days
REQUEST_TIMEOUT_S = 15

# A news item is flagged as a likely new grant if its title contains any of
# these tokens AND any of the sustainability/biopolymer relevance keywords below.
GRANT_TOKENS = [
    "grant", "scheme", "fund ", "funding", "challenge", "prize",
    "call for proposals", "applications open", "applications invited",
    "award", "innovation challenge", "launches", "announces",
    "non-dilutive", "incubator", "accelerator", "cohort",
]
RELEVANCE_TOKENS = [
    "sustainab", "compost", "biodegrad", "biopolymer", "bioplastic",
    "circular", "plastic", "packaging", "bio-econom", "biomanufactur",
    "climate", "cleantech", "green",
]

def is_likely_grant(title):
    t = title.lower()
    return any(g in t for g in GRANT_TOKENS) and any(r in t for r in RELEVANCE_TOKENS)

# ---------------------------------------------------------------
# Verified-sources whitelist
# ---------------------------------------------------------------
# Google News aggregates from thousands of sites including SEO blogs and
# unverified republishers. We only keep items whose `source` field matches a
# known editorial publisher OR whose URL is on a .gov.in / .nic.in domain.
# Match is case-insensitive substring against the source name.
TRUSTED_SOURCE_PATTERNS = [
    # --- Indian business / financial press ---
    "economic times", "livemint", "mint",
    "business standard", "financial express",
    "businessline", "the hindu businessline", "hindu businessline",
    "moneycontrol", "bq prime", "bloombergquint",
    # --- Indian startup / VC press ---
    "inc42", "yourstory", "your story", "entrackr", "vccircle", "the ken",
    # --- Indian general (national) ---
    "the hindu", "hindustan times", "indian express", "times of india", "tribune india",
    # --- Indian government / official ---
    "pib", "press information bureau", "ministry of",
    # --- Global wire / financial ---
    "reuters", "bloomberg", "associated press", "ap news",
    "financial times", "wall street journal", "wsj", "the guardian",
    # --- Climate / sustainability press ---
    "mongabay", "down to earth", "downtoearth", "carbon brief", "climate home",
    # --- Industry trade press (relevant to Bambrew) ---
    "biospectrum", "pharmabiz", "packaging south asia",
    # --- Devex (international grants) ---
    "devex",
]

TRUSTED_URL_DOMAINS = [
    ".gov.in/", ".nic.in/", "pib.gov.in",
    # Already-good news domains that sometimes ship without a clean source name:
    "reuters.com", "bloomberg.com", "ft.com", "economictimes.indiatimes.com",
    "livemint.com", "business-standard.com", "thehindu.com",
    "inc42.com", "yourstory.com", "moneycontrol.com",
]

def is_trusted_source(source, url=""):
    """True if the item comes from a whitelisted publisher OR a govt domain."""
    s = (source or "").lower().strip()
    u = (url or "").lower().strip()
    # 1. Govt / known-good URL domain (high signal, low collision)
    if any(d in u for d in TRUSTED_URL_DOMAINS):
        return True
    # 2. Reject obvious junk (empty source, generic placeholder)
    if not s or s in ("google news", "news", "press release"):
        return False
    # 3. Source-name whitelist
    return any(p in s for p in TRUSTED_SOURCE_PATTERNS)

# Best-effort: map a keyword fragment to a grantId in the dashboard
# so the auto-fetched card gets a clickable "Open grant" button.
GRANT_MAP = {
    "BioE3":                 "bioe3",
    "BIRAC BIG":             "birac-big",
    "DBT BIRAC":             "bioe3",
    "CBE JU":                "cbe-ju",
    "Horizon Europe circular": "cbe-ju",
    "EU LIFE":               "life",
    "ELEVATE Karnataka":     "elevate",
    "ELEVATE NxT":           "elevate-nxt",
    "NABARD climate":        "nabard-ngif",
    "NABARD Green Impact":   "nabard-ngif",
    "Green Climate Fund":    "gcf",
    "C-CAMP":                "ccamp-coe",
    "IKEA Foundation":       "ikea-ceic",
    "Bezos Earth Fund":      "bezos",
    "Innovate UK SSPP":      "sspp",
    "Earthshot Prize":       "earthshot",
    "NITI Aayog":            "anic",
    "Atal New India":        "anic",
    "Marico Innovation":     "marico",
    "TDB India-Finland":     "tdb-finland",
    "ADB Ventures":          "adb",
}

# ---------------------------------------------------------------
# Logging
# ---------------------------------------------------------------
logging.basicConfig(
    filename=str(LOG_FILE),
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
)
log = logging.getLogger("refresher")

# ---------------------------------------------------------------
# Fetch + parse
# ---------------------------------------------------------------
def fetch_keyword(kw):
    url = "https://news.google.com/rss/search?" + urlencode({
        "q": kw, "hl": "en-IN", "gl": "IN", "ceid": "IN:en"
    })
    req = Request(url, headers={"User-Agent": "Mozilla/5.0 BambrewNewsRefresher"})
    data = None
    for ctx in (_VERIFIED_CTX, _UNVERIFIED_CTX):
        try:
            data = urlopen(req, timeout=REQUEST_TIMEOUT_S, context=ctx).read()
            break
        except URLError as e:
            if "CERTIFICATE_VERIFY_FAILED" in str(e):
                continue
            log.warning("Fetch failed for %r: %s", kw, e)
            return []
        except Exception as e:
            log.warning("Fetch failed for %r: %s", kw, e)
            return []
    if data is None:
        log.warning("Fetch failed for %r: SSL trust path exhausted", kw)
        return []
    try:
        root = ET.fromstring(data)
    except ET.ParseError as e:
        log.warning("Parse failed for %r: %s", kw, e)
        return []

    items = []
    now_utc = datetime.datetime.now(datetime.timezone.utc)
    for item in root.findall(".//item"):
        title_raw = (item.findtext("title") or "").strip()
        link      = (item.findtext("link") or "").strip()
        pub_raw   = (item.findtext("pubDate") or "").strip()
        source_el = item.find("source")
        source    = source_el.text.strip() if (source_el is not None and source_el.text) else ""

        # Google News titles often look like "Headline - Source Name"
        if " - " in title_raw and not source:
            title, trailing = title_raw.rsplit(" - ", 1)
            source = trailing
        else:
            title = title_raw

        title = html.unescape(title).strip()
        source = html.unescape(source).strip() or "Google News"

        try:
            dt = parsedate_to_datetime(pub_raw) if pub_raw else None
        except Exception:
            dt = None
        if not dt:
            continue
        if (now_utc - dt).days > DAYS_BACK:
            continue

        items.append({
            "date":   dt.strftime("%Y-%m-%d"),
            "title":  title,
            "url":    link,
            "source": source,
            "matched_kw": kw,
            "_dt":    dt,
        })
    return items

def _norm_title(t):
    # strip the trailing " - Source Name" if present (Google News appends it)
    base = t.rsplit(" - ", 1)[0] if " - " in t else t
    return " ".join(base.lower().split())

def dedupe(items):
    """Dedupe by URL AND by normalised title. Google News rewrites URLs per
    search, so the same story matched by two keywords ends up with two
    different URLs but the same title."""
    seen_urls, seen_titles, out = set(), set(), []
    for it in items:
        norm_title = _norm_title(it["title"])
        if it["url"] in seen_urls or norm_title in seen_titles:
            continue
        seen_urls.add(it["url"])
        seen_titles.add(norm_title)
        out.append(it)
    return out

def grant_id_for(matched_kw, title):
    text = (matched_kw + " " + title).lower()
    for kw_frag, gid in GRANT_MAP.items():
        if kw_frag.lower() in text:
            return gid
    return None

# ---------------------------------------------------------------
# Emit JS
# ---------------------------------------------------------------
def js_str(s):
    if s is None:
        return ""
    return (
        s.replace("\\", "\\\\")
         .replace("'", "\\'")
         .replace("\n", " ")
         .replace("\r", " ")
         .strip()
    )

def guess_region(title, source):
    """Best-effort region guess from text."""
    text = (title + " " + source).lower()
    if any(k in text for k in ["india", "indian", "bharat", "delhi", "mumbai", "bengaluru", "karnataka"]): return "india"
    if any(k in text for k in ["eu ", " eu", "horizon europe", "european", "brussels", "cbe ju", "life programme"]): return "eu"
    if any(k in text for k in ["uk ", "united kingdom", "innovate uk", "british"]): return "uk"
    if any(k in text for k in ["world bank", "ifc", "adb", "undp", "unep", "gef", "gcf"]): return "multilateral"
    return "global"

def title_to_clean_name(title):
    # strip trailing " - Source" suffix
    base = title.rsplit(" - ", 1)[0] if " - " in title else title
    # truncate very long titles
    if len(base) > 110: base = base[:107] + "..."
    return base

def to_news_js(items):
    now_iso = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
    today_iso = datetime.datetime.now().strftime("%Y-%m-%d")

    lines = [
        f"const LAST_REFRESH = '{now_iso}';",
        "const NEWS = ["
    ]
    for it in items[:MAX_ITEMS_TOTAL]:
        gid = grant_id_for(it.get("matched_kw", ""), it["title"])
        sig = "Auto-fetched · Keyword: " + it.get("matched_kw", "")
        is_grant = "true" if is_likely_grant(it["title"]) else "false"
        lines.append(
            "  { date: '"        + js_str(it["date"])   + "', "
            "title: '"           + js_str(it["title"])  + "', "
            "grantId: "          + (("'" + gid + "'") if gid else "null") + ", "
            "significance: '"    + js_str(sig)          + "', "
            "source: '"          + js_str(it["source"]) + "', "
            "url: '"             + js_str(it["url"])    + "', "
            "isLikelyGrant: "    + is_grant             + " },"
        )
    lines.append("];")

    # AUTO_DISCOVERED_GRANTS: items flagged as likely grants get appended to
    # GRANTS via allGrants() in the dashboard so they appear automatically.
    lines.append("// __AUTO_GRANTS_START__")
    lines.append("const AUTO_DISCOVERED_GRANTS = [")
    auto_count = 0
    for it in items:
        if not is_likely_grant(it["title"]):
            continue
        # skip if already linked to a known grant
        if grant_id_for(it.get("matched_kw", ""), it["title"]):
            continue
        slug = "auto-" + str(abs(hash(it["url"])) % (10**8))
        name = title_to_clean_name(it["title"])
        region = guess_region(it["title"], it.get("source", ""))
        lines.append(
            "  { id: '"       + slug                + "', "
            "name: '"         + js_str(name)        + "', "
            "region: '"       + region              + "', "
            "type: 'deadline', "
            "value: 'See source', "
            "domain: 'Auto-detected from news headline', "
            "eligibility: 'Visit source link to verify eligibility', "
            "url: '"          + js_str(it["url"])   + "', "
            "notes: 'Auto-discovered from news on " + today_iso + " · keyword: " + js_str(it.get("matched_kw", "")) + "', "
            "discoveredOn: '" + today_iso           + "', "
            "autoDiscovered: true },"
        )
        auto_count += 1
        if auto_count >= 8:   # cap auto-discovered per run
            break
    lines.append("];")
    lines.append("// __AUTO_GRANTS_END__")
    return "\n".join(lines)

# ---------------------------------------------------------------
# Surgical rewrite of the dashboard
# ---------------------------------------------------------------
def update_dashboard(news_js):
    text = DASHBOARD.read_text(encoding="utf-8")
    start_marker = "// __NEWS_START__"
    end_marker   = "// __NEWS_END__"
    if start_marker not in text or end_marker not in text:
        log.error("Sentinels missing in dashboard HTML — aborting")
        return False
    start = text.index(start_marker) + len(start_marker)
    end   = text.index(end_marker)
    new_text = text[:start] + "\n" + news_js + "\n" + text[end:]

    backup = DASHBOARD.with_suffix(".html.bak")
    backup.write_text(text, encoding="utf-8")
    DASHBOARD.write_text(new_text, encoding="utf-8")
    log.info("Dashboard rewritten · backup at %s", backup)
    return True

# ---------------------------------------------------------------
# Main
# ---------------------------------------------------------------
def main():
    log.info("=== Refresh start ===")
    if not DASHBOARD.exists():
        log.error("Dashboard not found at %s — aborting", DASHBOARD)
        print("ERR · Dashboard not found at", DASHBOARD)
        sys.exit(1)
    if not KEYWORDS_FILE.exists():
        log.error("keywords.txt missing — aborting")
        print("ERR · keywords.txt missing")
        sys.exit(1)

    keywords = [
        line.strip()
        for line in KEYWORDS_FILE.read_text(encoding="utf-8").splitlines()
        if line.strip() and not line.strip().startswith("#")
    ]
    log.info("Loaded %d keywords", len(keywords))

    all_items = []
    for kw in keywords:
        items = fetch_keyword(kw)
        items = items[:PER_KEYWORD_MAX]
        log.info("  %r -> %d items", kw, len(items))
        all_items.extend(items)

    all_items = dedupe(all_items)
    before_trust = len(all_items)
    rejected = [it for it in all_items if not is_trusted_source(it["source"], it["url"])]
    all_items = [it for it in all_items if is_trusted_source(it["source"], it["url"])]
    log.info("Trusted-source filter: kept %d of %d items (dropped %d untrusted)",
             len(all_items), before_trust, len(rejected))
    for r in rejected[:10]:
        log.info("  dropped (untrusted source=%r): %s", r["source"], r["title"][:100])
    all_items.sort(
        key=lambda x: x["_dt"] or datetime.datetime.min.replace(tzinfo=datetime.timezone.utc),
        reverse=True,
    )
    kept = min(len(all_items), MAX_ITEMS_TOTAL)
    log.info("Total after dedupe + trust filter: %d; keeping top %d", len(all_items), kept)

    if not all_items:
        log.warning("Zero items fetched — leaving dashboard untouched")
        print("WARN · zero items fetched; dashboard unchanged")
        sys.exit(0)

    if update_dashboard(to_news_js(all_items)):
        print("OK · {} items written to NEWS array".format(kept))
        log.info("=== Refresh done (OK) ===\n")
    else:
        print("ERR · dashboard rewrite failed; see refresh.log")
        log.error("=== Refresh done (FAIL) ===\n")
        sys.exit(1)

def health_check():
    """Run a self-test — no writes, no network reach to the dashboard.
    Confirms: dashboard exists, sentinels present, keywords file readable,
    log writable, Python+SSL working, Google News reachable for one query."""
    ok = True
    print("Bambrew news refresher · health check")
    print("-" * 50)

    if DASHBOARD.exists():
        print(f"✓ dashboard found at {DASHBOARD}")
        text = DASHBOARD.read_text(encoding="utf-8")
        if "// __NEWS_START__" in text and "// __NEWS_END__" in text:
            print("✓ sentinels (__NEWS_START__ / __NEWS_END__) found in dashboard")
        else:
            print("✗ sentinels MISSING in dashboard — script can't write to it")
            ok = False
    else:
        print(f"✗ dashboard NOT found at {DASHBOARD}")
        ok = False

    if KEYWORDS_FILE.exists():
        kws = [l for l in KEYWORDS_FILE.read_text(encoding="utf-8").splitlines()
               if l.strip() and not l.strip().startswith("#")]
        print(f"✓ keywords.txt readable · {len(kws)} active keywords")
    else:
        print("✗ keywords.txt MISSING")
        ok = False

    try:
        LOG_FILE.touch(exist_ok=True)
        print(f"✓ log file writable: {LOG_FILE}")
    except Exception as e:
        print(f"✗ log file NOT writable: {e}")
        ok = False

    print("→ testing Google News fetch (1 generic query)…")
    test_items = fetch_keyword("India news")  # broad, virtually always returns results
    if test_items:
        print(f"✓ Google News reachable · {len(test_items)} test items returned")
    else:
        print("✗ Google News fetch returned 0 items — network or RSS broken?")
        ok = False

    plist_dest = pathlib.Path.home() / "Library/LaunchAgents/com.bambrew.newsrefresher.plist"
    if plist_dest.exists():
        print(f"✓ launchd schedule installed at {plist_dest}")
    else:
        print(f"⚠ launchd schedule not installed (run: bash install.sh)")

    print("-" * 50)
    print("RESULT:", "ALL CHECKS PASSED ✓" if ok else "ISSUES FOUND ✗ — see lines above")
    return 0 if ok else 1

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] in ("--health-check", "--test", "-t"):
        sys.exit(health_check())
    main()

File: api/refresh.js
// Vercel cron: runs every Monday at 9am IST (3:30 UTC)
// Fetches Google News for grant-related keywords, saves to KV.
// No GitHub push / redeploy needed — frontend reads from /api/auto-grants.
// KEYWORDS mirrors scripts/keywords.txt (the news-drawer pipeline's source list) —
// keep the two in sync; add new terms to keywords.txt and copy them here too.

import { kv } from '@vercel/kv';

const KEYWORDS = [
  'Bambrew sustainable packaging',
  'biopolymer packaging India',
  'PBAT compostable India',
  'PLA biodegradable India',
  '"BIRAC BioE3" 2026',
  '"BIRAC BIG" grant',
  '"DBT BIRAC" call 2026',
  '"NABARD" climate fund 2026',
  '"NABARD Green Impact Fund"',
  '"NITI Aayog" "Atal Innovation Mission" circular',
  '"Atal New India Challenge" startup',
  '"ELEVATE Karnataka" 2026',
  '"ELEVATE NxT" Karnataka startup',
  '"C-CAMP" sustainability grant',
  '"TDB" "India-Finland" 2026',
  '"Marico Innovation Foundation"',
  '"CBE JU" films coatings 2026',
  '"EU LIFE Programme" circular economy 2026',
  '"Horizon Europe" circular packaging',
  '"Innovate UK" SSPP plastic packaging',
  '"IKEA Foundation" India circular',
  '"Bezos Earth Fund" biopolymer',
  '"Earthshot Prize" 2027',
  '"Green Climate Fund" India NABARD',
  '"ADB Ventures" climate India',
  'India plastic waste management rules 2026',
  'India circular economy grant 2026',
  'sustainable packaging India funding 2026',
  'compostable packaging startup India funding',
  '"sustainable packaging" India 2026',
  '"compostable packaging" India',
  '"biodegradable packaging" India launch',
  '"plastic-free packaging" India',
  '"circular packaging" India brand',
  '"EPR" "plastic" India 2026',
  '"single-use plastic" ban India 2026',
  '"compostable bags" India launch',
  '"bamboo packaging" India',
  '"commits to" "plastic-free" packaging',
  '"pledges" "sustainable packaging"',
  '"100% sustainable packaging"',
  '"plastic-free pledge" brand',
  '"transition to" "compostable packaging"',
  '"phase out plastic" packaging brand',
  '"net zero" packaging India',
  '"ESG" packaging India brand',
  '"recycled content" packaging India brand',
  'FMCG "sustainable packaging" India',
  'D2C brand "plastic-free" India',
  '"compostable films" India',
  '"compostable film" launch',
  '"bio films" packaging India',
  '"compostable wrappers"',
  '"compostable mailers" India',
  '"biodegradable courier bags" India',
  '"compostable cutlery" India',
  '"biodegradable mulch film"',
  '"compostable food packaging" India',
  '"compostable trays" India',
  '"sustainability grant" India 2026',
  '"sustainability funding" startup India',
  '"climate tech funding" India 2026',
  '"cleantech grants" India',
  '"green tech grants" India',
  '"ESG funding" startup India',
  '"impact investment" circular economy India',
  '"carbon credit" startup grant India',
  '"climate fund" India 2026 startup',
  '"green finance" India startup 2026',
  '"environmental grants" India 2026',
  '"Extended Producer Responsibility" India',
  '"EPR" packaging India 2026',
  '"single-use plastic alternatives" India',
  '"plastic ban" India 2026 alternatives',
  '"ASEAN" circular economy grant',
  '"UNDP India" sustainability grant',
  '"UNEP India" plastic grant',
  '"World Bank India" sustainable packaging',
  '"compostable garbage bags" India',
  '"compostable bin liners" India',
  '"home compostable bags" launch',
  '"biodegradable trash bags" India',
  '"compostable waste bags" India',
  '"sustainability initiative" packaging India',
  '"launches" "sustainable packaging" India',
  '"launches" "compostable" India',
  '"introduces" "compostable packaging" India',
  '"rolls out" "sustainable packaging"',
  '"adopts" "compostable packaging"',
  '"switches to" "sustainable packaging" India',
  '"sustainability commitment" brand India',
  '"partners with" "sustainable packaging" India',
  '"collaborates" "sustainable packaging" India',
  'brand "goes plastic-free" India',
  'company "eco-friendly packaging" India launch',
  '"sustainable packaging" "Series A" OR "seed" India 2026',
  '"green bond" India startup 2026',
  '"climate tech" India funding 2026'
];

const GRANT_TOKENS = ['grant','scheme','fund ','funding','challenge','prize','call for proposals',
  'applications open','applications invited','award','innovation challenge','launches','announces',
  'non-dilutive','incubator','accelerator','cohort'];
const RELEVANCE_TOKENS = ['sustainab','compost','biodegrad','biopolymer','bioplastic','circular',
  'plastic','packaging','bio-econom','biomanufactur','climate','cleantech','green'];
const TRUSTED = ['economic times','business standard','mint','livemint','financial express',
  'businessline','moneycontrol','ndtv','bloomberg','reuters','times of india','india today',
  'inc42','yourstory','entrackr','vccircle','techcrunch','fortune india','business today',
  'down to earth','mongabay','eco-business','pib','press information bureau','pti','ani',
  'biospectrum','biovoice','pharmabiz','expresspharma','dbt.gov','birac.nic','dst.gov',
  'nabard.org','niti.gov','startupindia.gov','makeinindia.com','abi.org.in'];

function isGrant(title) {
  const t = title.toLowerCase();
  return GRANT_TOKENS.some(g => t.includes(g)) && RELEVANCE_TOKENS.some(r => t.includes(r));
}
function isTrusted(source) {
  if (!source) return false;
  const s = source.toLowerCase();
  return TRUSTED.some(t => s.includes(t));
}
function makeId(url) {
  let h = 0;
  for (let i = 0; i < url.length; i++) h = (Math.imul(31, h) + url.charCodeAt(i)) | 0;
  return 'auto-' + Math.abs(h).toString(36);
}
function parseRSS(xml) {
  const items = [];
  const re = /<item>([\s\S]*?)<\/item>/g;
  let m;
  while ((m = re.exec(xml))) {
    const b = m[1];
    const get = tag => {
      const r = new RegExp(`<${tag}[^>]*><!\\[CDATA\\[([\\s\\S]*?)\\]\\]><\\/${tag}>|<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, 'i');
      const x = r.exec(b);
      return x ? (x[1] || x[2] || '').trim() : '';
    };
    const decode = s => s.replace(/&amp;/g,'&').replace(/&lt;/g,'<').replace(/&gt;/g,'>').replace(/&#39;/g,"'").replace(/&quot;/g,'"');
    const title = decode(get('title')), link = get('link'), pubDate = get('pubDate'), source = decode(get('source'));
    if (title && link) items.push({ title, link, pubDate, source });
  }
  return items;
}
async function fetchKeyword(kw) {
  const url = `https://news.google.com/rss/search?q=${encodeURIComponent(kw)}&hl=en-IN&gl=IN&ceid=IN:en`;
  try {
    const r = await fetch(url, { headers: { 'User-Agent': 'Mozilla/5.0' }, signal: AbortSignal.timeout(8000) });
    if (!r.ok) return [];
    return parseRSS(await r.text());
  } catch { return []; }
}

export default async function handler(req, res) {
  try {
    const cutoff = Date.now() - 30 * 24 * 60 * 60 * 1000;
    const results = [];
    for (let i = 0; i < KEYWORDS.length; i += 10) {
      const items = await Promise.all(KEYWORDS.slice(i, i + 10).map(fetchKeyword));
      results.push(...items.flat());
    }

    const seen = new Set();
    const unique = results.filter(x => { if (seen.has(x.link)) return false; seen.add(x.link); return true; });

    const filtered = unique.filter(x => {
      if (!isTrusted(x.source)) return false;
      if (x.pubDate && new Date(x.pubDate).getTime() < cutoff) return false;
      return isGrant(x.title);
    });

    filtered.sort((a, b) => new Date(b.pubDate || 0) - new Date(a.pubDate || 0));
    const today = new Date().toISOString().slice(0, 10);

    const grants = filtered.slice(0, 12).map(x => ({
      id: makeId(x.link),
      name: x.title,
      region: 'india',
      category: /award|prize/i.test(x.title) ? 'award' : 'grant',
      type: 'deadline',
      value: 'See source',
      domain: 'Auto-detected from news headline',
      eligibility: 'Visit source link to verify eligibility',
      url: x.link,
      notes: `Auto-discovered from news on ${x.pubDate ? new Date(x.pubDate).toISOString().slice(0,10) : today}`,
      discoveredOn: x.pubDate ? new Date(x.pubDate).toISOString().slice(0, 10) : today,
      autoDiscovered: true,
    }));

    await kv.set('auto-grants', { grants, updatedAt: today });

    return res.json({ ok: true, count: grants.length, date: today });
  } catch (e) {
    return res.status(500).json({ error: e.message });
  }
}
